src/Adapter/CustomerService/CommandHandler/ReplyToCustomerThreadHandler.php

```diff
@@ -0,0 +1,174 @@
+use PrestaShop\PrestaShop\Core\Domain\CustomerService\ValueObject\CustomerThreadStatus;
+use ShopUrl;
+use Symfony\Component\Translation\TranslatorInterface;
+use Tools;
+use Validate;
+
+/**
+ * @internal
+ */
+final class ReplyToCustomerThreadHandler implements ReplyToCustomerThreadHandlerInterface
+{
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @param Context $context
+     */
+    public function __construct(Context $context)
+    {
+        $this->context = $context;
+        $this->translator = $context->getTranslator();
+    }
+
+    /**
+     * @param ReplyToCustomerThreadCommand $command
+     */
+    public function handle(ReplyToCustomerThreadCommand $command)
+    {
+        $customerThread = new CustomerThread(
+            $command->getCustomerThreadId()->getValue()
+        );
+
+        ShopUrl::cacheMainDomainForShop((int) $customerThread->id_shop);
+
+        $this->createCustomerMessage(
+            $customerThread,
+            $command->getReplyMessage()
+        );
+
+        $replyWasSent = $this->sendReplyEmail($customerThread, $command);
+
+        if ($replyWasSent) {
+            $customerThread->status = CustomerThreadStatus::CLOSED;
+            $customerThread->update();
+        }
+    }
+
+    /**
+     * @param CustomerThread $customerThread
+     * @param string $replyMessage
+     *
+     * @return CustomerMessage
+     */
+    private function createCustomerMessage(CustomerThread $customerThread, $replyMessage)
+    {
+        $customerMessage = new CustomerMessage();
+        $customerMessage->id_employee = (int) $this->context->employee->id;
+        $customerMessage->id_customer_thread = $customerThread->id;
+        $customerMessage->ip_address = (int) ip2long(Tools::getRemoteAddr());
+        $customerMessage->message = $replyMessage;
+
+        if (false === $customerMessage->validateField('message', $customerMessage->message)) {
+            throw new CustomerServiceException('Invalid reply message');
+        }
+
+        if (false === $customerMessage->add()) {
+            throw new CustomerServiceException('Failed to add customer message');
+        }
+
+        return $customerMessage;
+    }
+
+    /**
+     * @param CustomerThread $customerThread
+     * @param ReplyToCustomerThreadCommand $command
+     *
+     * @return bool
+     */
+    private function sendReplyEmail(CustomerThread $customerThread, ReplyToCustomerThreadCommand $command)
+    {
+        $customer = new Customer($customerThread->id_customer);
+
+        $params = [
+            '{reply}' => Tools::nl2br($command->getReplyMessage()),
+            '{link}' => Tools::url(
+                $this->context->link->getPageLink('contact', true, null, null, false, $customerThread->id_shop),
+                'id_customer_thread=' . (int) $customerThread->id . '&token=' . $customerThread->token
```

Comment:
😥 
Maybe we should enhance `getPageLink()` function to be able to pass it parameters

Reply:
Not sure, `getPageLink` already has 7 arguments, we would end up with something like `getPageLink('my-page', null, null, null, false, null, false, ['my_param' => 1])`. :/

Reply:
What about something like `getPageLink(...)->addParameters(...)` ?

Reply:
Hmmm, I think that would be a huge BC break, as `getPageLink(): string` returns string and now it should return some object to be able to chain on it, this is how I understand your example, is it right?

Reply:
If you take it literally yes 😄 
But we can make it happen by creating a new function `getPageLinkBuilder()` that returns a BuildableLink on which we can call `addParameters()`

And `getPageLinkBuilder()` accepts the same arguments that `getPageLink()` accepts, and `getPageLinkBuilder()->buildLink()` returns the same thing as`getPageLink()`

Reply:
So complex to get a link, module devs will explode. :boom:  Either way, if we decide to do so, it should be part of separate PR.